### PR TITLE
NetworkChannel update for custom messages

### PR DESCRIPTION
**Pull request by Wmaxlees**
Updated the NetworkChannel class to allow modders to add new NetworkChannels externally. Also, changed a field in the Crafting GUI open message to be protected so any subclasses can access this.

# Changes proposed in this pull request:
- Make NetworkChannel.registerMessage public
- Change a field in the Crafting GUI message to protected so it can be subclassed effectively.

[x] Yes I tested this before submitting it.
[x] I also did a multiplayer test.

Review please

### CODE REVIEW

**Comment by Thodor12** [comment]
Before we go ahead and merge this, could you give an example of a place in the code where you require this?

File: src/main/java/com/minecolonies/core/network/messages/server/colony/building/OpenCraftingGUIMessage.java [hunk 1]
```diff
@@ -29,7 +29,7 @@ public class OpenCraftingGUIMessage extends AbstractBuildingServerMessage<IBuild
     /**
      * The type of container.
      */
-    private int id;
+    protected int id;
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/core/network/messages/server/colony/building/OpenCraftingGUIMessage.java:32]: Is this change strictly needed?
You can just create a custom message class and use that instead of modifying and extending this one.

**Comment by uecasm** [comment]
I already mentioned in Discord that this sort of thing is counterproductive.  If you want to make custom messages, make your own channel for it, using standard Forge code.

You are **absolutely forbidden** by how Forge networking works from adding additional messages into some other mod's channel.

There's no reason you can't override the GUI view and send your own message when the button is clicked.  Or alternatively, the `CraftingType` could be extended with a method to open the appropriate GUI view, though that's trickier because GUIs are client-side-only and you can trigger classloading problems if you're not careful.

**Comment by Thodor12** [comment]
> I already mentioned in Discord that this sort of thing is counterproductive. If you want to make custom messages, make your own channel for it, using standard Forge code.
> 
> You are **absolutely forbidden** by how Forge networking works from adding additional messages into some other mod's channel.
> 
> There's no reason you can't override the GUI view and send your own message when the button is clicked. Or alternatively, the `CraftingType` could be extended with a method to open the appropriate GUI view, though that's trickier because GUIs are client-side-only and you can trigger classloading problems if you're not careful.



> I already mentioned in Discord that this sort of thing is counterproductive. If you want to make custom messages, make your own channel for it, using standard Forge code.
> 
> You are **absolutely forbidden** by how Forge networking works from adding additional messages into some other mod's channel.
> 
> There's no reason you can't override the GUI view and send your own message when the button is clicked. Or alternatively, the `CraftingType` could be extended with a method to open the appropriate GUI view, though that's trickier because GUIs are client-side-only and you can trigger classloading problems if you're not careful.

I mean, realistically if the problem is the OpenCraftingGUIMessage, you just don't use said message.
That message is only implemented to work for smelting, recipes and brewing.

Considering you have to make your own module view for your custom crafting module, you also have to implement a custom message to open a separate view to add a message, and so forth.

**Comment by marchermans** [comment]
I am 100% against this. Our network is our protected namespace. If you need your own payloads, configure your own channel under your own network.

**Review by marchermans — CHANGES_REQUESTED** [review]
This will not be merged if it were up to me as it violates several design objectives, language objectives and forge design and architecture objectives

**Comment by Wmaxlees** [comment]
Understood. I'll close the pull request. Is there a proper channel in the future for me to discuss changes with the dev team before making a pull request? I'd like to save everyone's time if the idea is a no-go. Particularly for bigger changes like converting the IToolType system into a registry (which was also suggested on discord as a PR).
